redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt

```diff
@@ -28,128 +28,284 @@ import app.cash.redwood.lazylayout.widget.RefreshableLazyList
+//        val item = items.itemForGlobalIndex(sizeForItemAtIndexPath.item.toInt())
+//
+//        if (collectionViewFlowLayout.scrollDirection == UICollectionViewScrollDirection.UICollectionViewScrollDirectionVertical) {
+//          return CGSizeMake(collectionView.frame().useContents { size.width }, item.value.intrinsicContentSize.useContents { height })
+//        } else {
+//          return CGSizeMake(item.value.intrinsicContentSize.useContents { width }, collectionView.frame().useContents { size.height })
+//        }
+      }
+
+      override fun scrollViewDidScroll(scrollView: UIScrollView) {
+        val visibleIndexPaths = collectionView.indexPathsForVisibleItems()
+
+        if (visibleIndexPaths.isNotEmpty()) {
+          // TODO: Optimize this
+          viewPortListCoordinator.updateViewport(
+            visibleIndexPaths.minOf { (it as NSIndexPath).item.toInt() },
+            visibleIndexPaths.maxOf { (it as NSIndexPath).item.toInt() }
+          )
+          if (!items.isFinishedInitialLoad) {
+            items.isFinishedInitialLoad = true
+            println("🟢 isFinishedInitialLoad scrollViewDidScroll")
+          }
+        }
+      }
     }
 
-  internal val tableView = UITableView()
-    .apply {
-      dataSource = tableViewDataSource
-      delegate = tableViewDelegate
-      rowHeight = 64.0 // TODO: size rows by their content.
+  internal val collectionView = UICollectionView(
+    frame = CGRectZero.readValue(),
+    collectionViewLayout = this.collectionViewFlowLayout
+  ).apply {
+      dataSource = collectionViewDataSource
+      delegate = collectionViewDelegate
       prefetchingEnabled = false
+
       registerClass(
-        Cell(UITableViewCellStyleDefault, reuseIdentifier).classForCoder() as ObjCClass?,
-        forCellReuseIdentifier = reuseIdentifier,
+        LazyListContainerCell(CGRectZero.readValue()).classForCoder() as ObjCClass?,
+        reuseIdentifier
       )
     }
 
-  private lateinit var onViewportChanged: (firstVisibleItemIndex: Int, lastVisibleItemIndex: Int) -> Unit
-  private var itemsBefore = 0
-  private var itemsAfter = 0
-
+  // LazyList methods
   override fun isVertical(isVertical: Boolean) {
+    println("🟤 isVertical")
     if (!isVertical) {
-      // TODO UITableView only supports vertical scrolling. Switch to UICollectionView.
-      TODO()
+      collectionViewFlowLayout.scrollDirection = if (isVertical) {
+        UICollectionViewScrollDirection.UICollectionViewScrollDirectionVertical
+      } else {
+        UICollectionViewScrollDirection.UICollectionViewScrollDirectionHorizontal
+      }
     }
   }
 
   override fun onViewportChanged(onViewportChanged: (firstVisibleItemIndex: Int, lastVisibleItemIndex: Int) -> Unit) {
+    println("🟤 onViewportChanged")
     this.onViewportChanged = onViewportChanged
   }
 
   override fun itemsBefore(itemsBefore: Int) {
-    this.itemsBefore = itemsBefore
+    println("🟤 itemsBefore $itemsBefore")
```

Comment:
I think something that we should consider is having a way for us to pass along a full transaction or model, instead of individual binding callbacks.

Reply:
The current design is imposed on us by the underlying Compose mechanism. We can't really change it because that would break direct usage (i.e., when not using Treehouse).

You can, however, implement `ChangeListener` on a widget type which gives you a callback after each batch of changes. This should effectively allow you to create the same thing. Instead of doing any action in these individual functions, instead, write the state to properties and then only act when that callback occurs.

Reply:
What @underscoretang said sounds similar to [what I said in the `ChangeListener` proposal issue](https://github.com/cashapp/redwood/issues/1133#issuecomment-1561666108). Considering all three change most of the time, what's the downside of that approach (apart from the class wrapper)?

Reply:
going to try using the changeListener approach here. the hope is to onlyReload and update the views when all the changes are committed, otherwise we end up with a bunch of intermediate updates that might not have all the state changes ready (esp on agressively scrolling)